---
title: "Literature Synthesis"
author: "Group 10 members"
date: "4/12/2017"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr);
library("stringr");
library(ggplot2);
library(dplyr);
library(tidyr);
library(RColorBrewer);
library(dendextend);
library(gplots);
library(cluster);
library(igraph)
library(ggraph)
library(gridExtra)
library(grid)
library(gProfileR) ## might need to be installed
#library(org.Sc.sgd.db)
```

# Question 1 - GO Analysis 

At first we wanted to familiarize ourselves with the database, shown below. 

```{r}
#columns(org.Sc.sgd.db)
#select(org.Sc.sgd.db, "YAL022C", "ALIAS")
```

### Gene Clustering GO Analysis 

Below we pulled in the microarray data and our k-means clustering analysis.

```{r}
kuang <- read.csv("./datasets/kuang-2014-microarray-expression.csv") ##local- needs to be changed 
kuang_t <- as.data.frame(t(kuang[,-1]))
n <- kuang$Gene.name # first remember the names
colnames(kuang_t) <- n
kuang_t.cor <- kuang_t %>% cor(use="pairwise.complete.obs")  ##generate correlation measurements
kuang.dist <- as.dist(1 - kuang_t.cor)  ##use correlation to find distance measurements 
kuang.tree <- hclust(kuang.dist, method="complete")
kuang.dend <- as.dendrogram(kuang.tree) ## created & priviewed dendrogram object
k=3
kuang.kmedoids <- pam(kuang.dist, k) # create k-medoids clustering with k clusters
kclusters <- kuang.kmedoids$cluster
kclusters <- kclusters[order.dendrogram(kuang.dend)]
```

```{r}
kcluster1.genes <- names(kclusters[kclusters == 1])
kcluster2.genes <- names(kclusters[kclusters == 2])
kcluster3.genes <- names(kclusters[kclusters == 3])
length(kcluster1.genes)
length(kcluster2.genes)
length(kcluster3.genes)
```

In our k-means clustering we only created 3 clusters. We did this to be consistent with the paper. This is described in more detail in our clustering homework. However, in this homework, we just wanted to convey the sizes of our k-means clusters. These clusters are quite large and we decided this was not ideal for GO analysis. We feared that large cluster sizes could lead to clusters with a wide range of functions, which we wanted to avoid for this particular analysis. For this reason, as an alternative, we carry out GO analysis with our hierarchical clusters from the RNASeq data.

#RnaSeq

Below we pulled in the RNASeq data and perform the hierarchical clustering as done in the clustering homework.

```{r}

rnaseq <- read.csv("./datasets/kuang-2014-rnaseq-expression.csv") #local needs to be changed
rnatime <- read.csv("./datasets/kuang-2014-rnaseq-expression-timepts.csv")
# first remember the names
nrna <- rnaseq$Gene.name

# transpose all but the first column (Gene.name)
rnaseq_t_zeros <- as.data.frame(t(rnaseq[,-1]))
colnames(rnaseq_t_zeros) <- nrna

rnaseq_t_zeros.time <- cbind(rnaseq_t_zeros, time = rnatime$hour)

rnaseq_t <- rnaseq_t_zeros[, colSums(rnaseq_t_zeros != 0) > 0]
rnaseq_t.time <- cbind(rnaseq_t, time = rnatime$hour)

rnaseq.cor <- cor(rnaseq_t, use = "pairwise.complete.obs") 
rnaseq.dist <- as.dist(1-rnaseq.cor)

rnaseq.tree <- hclust(rnaseq.dist, method = "complete") 
rnaseq.dend <- as.dendrogram(rnaseq.tree)

rnaseq.clusters <- cut(rnaseq.dend, h = 1.9)

rnaseq.clusters.cluster1 <- cut(rnaseq.clusters$lower[[1]], h = 1.5)
rnaseq.clusters.cluster1.clusterAgain <- cut(rnaseq.clusters.cluster1$lower[[1]], h = 1.2)

rnaseq.clusters.cluster2 <- cut(rnaseq.clusters$lower[[2]], h = 1.6)

rnaseq.clusters.cluster3 <- cut(rnaseq.clusters$lower[[3]], h = 1.3)
rnaseq.clusters.cluster3.cutAgain <- cut(rnaseq.clusters.cluster3$lower[[1]], h = .85)

rnaseq.clusters.cluster4 <- cut(rnaseq.clusters$lower[[4]], h = 1.48)
```

In our hierarchical clustering of the RNASeq data we arrived at 18 final clusters. Their names and sizes are listed below.

```{r}
rnaseq.clusters.cluster1.clusterAgain$lower[[1]]
rnaseq.clusters.cluster1.clusterAgain$lower[[2]]
rnaseq.clusters.cluster2$lower[[1]]
rnaseq.clusters.cluster2$lower[[2]]
rnaseq.clusters.cluster2$lower[[3]]
rnaseq.clusters.cluster2$lower[[4]]
rnaseq.clusters.cluster3.cutAgain$lower[[1]]
rnaseq.clusters.cluster3.cutAgain$lower[[2]]
rnaseq.clusters.cluster3.cutAgain$lower[[3]]
rnaseq.clusters.cluster3.cutAgain$lower[[4]]
rnaseq.clusters.cluster3.cutAgain$lower[[5]]
rnaseq.clusters.cluster3$lower[[2]]
rnaseq.clusters.cluster3$lower[[3]]
rnaseq.clusters.cluster4$lower[[1]]
rnaseq.clusters.cluster4$lower[[2]]
rnaseq.clusters.cluster4$lower[[3]]
rnaseq.clusters.cluster4$lower[[4]]
rnaseq.clusters.cluster4$lower[[5]]
```

*Largest*

From the original 18 clusters we chose the 3 largest to perform GO analysis, shown below.  

```{r}
rnaseq.clusters.cluster1.clusterAgain$lower[[1]]
rnaseq.clusters.cluster3.cutAgain$lower[[2]]
rnaseq.clusters.cluster4$lower[[4]]
```

*Ensure their expression varies* 

However, after we chose these clusters we wanted to varify that they contained genes that changed expression over time. In other words we did not want to run GO analysis on clusters with genes whose expression did not vary with time. A cluster with non-variant gene expression would imply that the cluster did not respond to the stress and thus its genes functions are not particularly relevant to the cell's response to the starvation stress during the metabolic process.

To varify variant expression we plotted the first 5 genes' expression, in each cluster, over time and observed the results.

```{r}
rnaseq_t_zeros.time.long <- gather(rnaseq_t_zeros.time, gene, expression, -time);
cols <- c("time", "gene", "expression");
rnaseq_t_zeros.time.long.filtered <- rnaseq_t_zeros.time.long[,cols];
```

```{r}
cluster1.genes.names <- as.character(labels(rnaseq.clusters.cluster1.clusterAgain$lower[[1]]))
cluster1.genes.first5 <- cluster1.genes.names[1:5]
rnaseq_t_zeros.time.long.filtered %>%
  filter(gene %in% cluster1.genes.first5) %>%
  ggplot(aes(x = time, y = expression, color = gene)) + geom_line()
```

```{r}
cluster2.genes.names <- as.character(labels(rnaseq.clusters.cluster3.cutAgain$lower[[2]]))
cluster2.genes.first5 <- cluster2.genes.names[1:5]
rnaseq_t_zeros.time.long.filtered %>%
  filter(gene %in% cluster2.genes.first5) %>%
  ggplot(aes(x = time, y = expression, color = gene)) + geom_line()
```

```{r}
cluster3.genes.names <- as.character(labels(rnaseq.clusters.cluster4$lower[[4]]))
cluster3.genes.first5 <- cluster3.genes.names[1:5]
rnaseq_t_zeros.time.long.filtered %>%
  filter(gene %in% cluster3.genes.first5) %>%
  ggplot(aes(x = time, y = expression, color = gene)) + geom_line()
```

From the plots above it was clear that all three clusters contained genes with variant expression over time. Thus, we decided to continue with these three clusters.

# GO for Hierarchical Clustering

In the following section we carry out the GO analysis for these three clusters.

**P-value cutoff**

At this point we did need to make a decision about what p-value cutoff to use. We used a cutoff of 0.001. It is important to note this cutoff is just a threshold for the *gprofiler* to return results, not a threshold for us to determine significance. Thus, we decided to chose a relatively lose cutoff value for the *gprofiler* and determine the significance of each returned turn retroactively. In this manner we would avoid missing significant terms. 

### Cluster 1

```{r}
gene.clu <- lapply(rnaseq.clusters.cluster1.clusterAgain$lower, labels)
cluster1.profile <- gprofiler(gene.clu[[1]], 
                       organism = "scerevisiae",
                       max_p_value = 0.001,  # set p-value cutoff
                       src_filter = "GO", # consider all processes
                       hier_filtering = "moderate"
                       )
cluster1.profile.sorted <- cluster1.profile[order(cluster1.profile$p.value),] ## sort based on p.value
columns.of.interest <- c(3,4,8,9,12)
head(cluster1.profile.sorted[,columns.of.interest]) ##make more readable
```

# GO Term Criteria

We had 3 main criteria for chosing which GO terms to investigate further
1. Significant p-values - we wanted to investigate meaningful (low p-value) GO terms
2. Specificity - we wanted to avoid investigating GO terms that were too broad. Since GO terms are organized in a tree structure, we wanted to avoid GO terms "too high up on the tree."
3. Relevance to paper - we wanted to investigate GO terms that were relevant to the metabolic process or chromatin states. Of course, we didn't want to limit ourselves from finding discoveries in other areas. However, we also didn't want to lose our *focus*. 

### Cluster 1 - GO Terms and papers 

For the first cluster, the gene ontology term with the smallest p-value is **monocarboxylic acid metabolic process**.  With a p-value of 5.94e-09, the probability of this term appearing by chances is almost impossible.  This is very encouraging because our assigned paper deals with chromatin states across metabolic processes.

Citation:
Kuang Z, Cai L, Zhang X, Ji H, Tu BP, Boeke JD. High-temporal-resolution view of transcription and chromatin states across distinct metabolic states in budding yeast. Nat Struct Mol Biol. 2014 Oct;21(10):854-63. doi: 10.1038/nsmb.2881. PubMed PMID: 25173176; PubMed Central PMCID: PMC4190017.

The next term of note that appeared in this cluster is **oxidation-reduction process**.  This is noteworthy because our assigned paper is analyzing the metabolic process in glucose-limited conditions, so to see a cluster mapped to a resource reduction makes sense. TODO check this. don't think it makes sense

Citation:
Le Fourn, C., Brasseur, G., Brochier-Armanet, C., Pieulle, L., Brioukhanov, A., Ollivier, B. and Dolla, A. (2011), An oxygen reduction chain in the hyperthermophilic anaerobe Thermotoga maritima highlights horizontal gene transfer between Thermococcales and Thermotogales. Environmental Microbiology, 13: 2132â€“2145. doi:10.1111/j.1462-2920.2011.02439.x

Piggybacking off of the first gene ontology term for this cluster, the final term of interest is **nicotinamide nucleotide metabolic process**.  This term is interesting for its low p-value, as well as its relation to the metabolic process, which is the process of focus of our paper.

Citation:
Foster JW, Moat AG. Nicotinamide adenine dinucleotide biosynthesis and pyridine nucleotide cycle metabolism in microbial systems. Microbiological Reviews. 1980;44(1):83-105.

### Cluster 2

```{r}
gene.clu2 <- lapply(rnaseq.clusters.cluster3.cutAgain$lower, labels)
cluster2.profile <- gprofiler(gene.clu2[[2]], 
                       organism = "scerevisiae",
                       max_p_value = 0.001,  # set p-value cutoff
                       src_filter = "GO", # consider all processes
                       hier_filtering = "moderate"
                       )
cluster2.profile.sorted <- cluster2.profile[order(cluster2.profile$p.value),] 
head(cluster2.profile.sorted[,columns.of.interest]) ##make more readable
```
There are multiple terms related to the cytoplasm: **cytoplasmic translation**, **cytosolic ribosome**, and **nucleocytoplasmic transport**.  Metabolic processes occur in the cytoplasm, so this lines up well to the other metabolic GO terms we have seen, as well as our paper.

Citation:
Tripathi A, Mandon EC, Gilmore R, Rapoport TA. Two alternative binding mechanisms connect the protein translocation Sec71/Sec72 complex with heat shock proteins. The Journal of Biological Chemistry. PMID 28286332 DOI: 10.1074/jbc.M116.761122 

With a p-value of 6.76e-72, another gene ontology term of note out of cluster 2 is **structural constituent of ribosome**.  The paper by Kuang et al., that is the basis for our project, deals with changes in chromatin status across robust metabolic cycles.  It is possibe that the structural consituent of ribosome plays a large role in these chromatic state changes.

Citation:
Chen K, Deng S, Lu H, Zheng Y, Yang G, Kim D, et al. (2013) RNA-Seq Characterization of Spinal Cord Injury Transcriptome in Acute/Subacute Phases: A Resource for Understanding the Pathology at the Systems Level. PLoS ONE 8(8): e72567. https://doi.org/10.1371/journal.pone.0072567

Similar to structural constituent of ribosome, the term **ribonucleoprotein complex biogenesis** also appeared for this second cluster with a p-value of 6.61e-34.  This is a process that results in the assembly and arrangement of consituent parts of a complex containing RNA and proteins.  Chromosomes consist of DNA, so there could be a connection there.

Citation:
Dragon, F. et al. A large nucleolar U3 ribonucleoprotein required for 18S ribosomal RNA biogenesis.
Nature 417, 967â€“970 (2002).

### Cluster 3

```{r}
gene.clu3 <- lapply(rnaseq.clusters.cluster4$lower, labels)
cluster3.profile <- gprofiler(gene.clu3[[4]], 
                       organism = "scerevisiae",
                       max_p_value = 0.001,  # set p-value cutoff
                       src_filter = "GO", # consider all processes
                       hier_filtering = "moderate"
                       )
cluster3.profile.sorted <- cluster3.profile[order(cluster3.profile$p.value),] 
head(cluster3.profile.sorted[,columns.of.interest]) ##make more readable
```

The first notable term, with the lowest p-value in this cluster, is **ubiquitin-dependent protein catabolic process**.  Catabolism is the set of metabolic pathways that breaks down molecules for energy purposes or for other anabolic reactions.

Citation:
Dong, Y., Hu, J., Fan, L., & Chen, Q. (2017). RNA-Seq-based transcriptomic and metabolomic analysis reveal stress responses and programmed cell death induced by acetic acid in Saccharomyces cerevisiae. Scientific Reports, 7, 42659. doi:10.1038/srep42659

The next lowest p-value in the third cluster is **vesicle-mediated transport**.  This term relates to the transportation of substances in membrane-bound vesicles.

Citation:
Vesicle-mediated protein transport: regulatory interactions between the Vps15 protein kinase and the Vps34 PtdIns 3-kinase essential for protein sorting to the vacuole in yeast. The Journal of Cell Biology. 1995;129(2):321-334.

With a p-value of 7.00e-09, **endomembrane system** is very interesting.  The endomembrane system regulates protein traffic and, more notably, performs metabolic functions in the cell.  Again, we are brougt back to the metabolic process - the focus of our paper.

Citation:
Hariri, H., Ugrankar, R., Liu, Y., & Henne, W. M. (2016). Inter-organelle ER-endolysosomal contact sites in metabolism and disease across evolution. Communicative & Integrative Biology, 9(3). doi:10.1080/19420889.2016.1156278

### Re-occuring Terms and Patterns

Metabolism, either directly or indirectly, appears in every cluster.  In the first cluster it is the GO term with the lowest p-value.  In the second cluster we see many terms dealing with the cytoplasm.  After doing more research we found that metabolic processes occur in the cytoplasm.  In the third cluster the GO term with the smallest p-value deals with catabolic processes, which is the phase of the metabolic process that distintegrates complex organic compounds to release energy.  It is encouraging to see metabolism being repeated in all three clusters because the paper our data comes from is researching chromatin states across metabolic cycles.  Therefore it makes sense that genes related to the metabolic process are expressed highly and in all the clusters.

There are no other terms that appear in multiple clusters.

#TF Binding Clusters

Now carry out GO analysis on clusters we identified in our TF network analyses. Here are steps we did to get our 3 TF clusters (same as in the TF homework):

```{r}

kuangtf <- read.csv("./datasets/kuang-2014-TFbinding.csv")
time <- read.csv("./datasets/kuang-2014-TFbinding-timepts.csv")
kuangrna <- read.csv("./datasets/kuang-2014-rnaseq-expression.csv")   ##used later - included in the import section of the code for clarity

Gcn <- kuangtf[,c(1:15)]
Esa <- kuangtf[,c(1,16:29)]
Set <- kuangtf[,c(1,30:43)]

Gcn.data <- as.data.frame(t(Gcn[,-1]))
colnames(Gcn.data) <- Gcn$Gene.ID
Gcn.time <- cbind(Gcn.data, time = time$hour)

Esa.data <- as.data.frame(t(Esa[,-1]))
colnames(Esa.data) <- Esa$Gene.ID
Esa.time <- cbind(Esa.data, time = time$hour)

Set.data <- as.data.frame(t(Set[,-1]))
colnames(Set.data) <- Set$Gene.ID
Set.time <- cbind(Set.data, time = time$hour)

#get the genes in the tf and rna data
TF.genes <- unique(kuangtf[,1])
rna.genes <- unique(kuangrna[,1])

#get rid of duplicated rows in the rna data
rna.unique <- unique(kuangrna)

#filter the rna data for the tf genes
rna.Genes <- filter(rna.unique, rna.unique[,1] %in% TF.genes)

#get the unique genes from this rna data
#may not need to do this again but did it just to be safe
rna.genes.u <- unique(rna.Genes[,1])

#It appears that all of the genes present in the TF data are not present in the rna data
#So need to filter again to get rid of these genes
kuangtf.new <- filter(kuangtf, TF.genes %in% rna.genes.u)

#now remove last two columns so the times add up
rna.new <- rna.Genes[,1:15]

dim(rna.new)
dim(kuangtf.new)

#now extract the different TFs
Gcn.new <- kuangtf.new[,c(1:15)]
dim(Gcn.new)
Esa.new <- kuangtf.new[,c(1,16:29)]
dim(Esa.new)
Set.new <- kuangtf.new[,c(1,30:43)]
dim(Set.new)

#here we see the 69 genes that are in the TF data but not the rnaseq
length(which( ! TF.genes %in% rna.genes))

#sort so that the genes are in the same order
Gcn.sorted <- Gcn.new[match(rna.new[,1], Gcn.new[,1]),]
Esa.sorted <- Esa.new[match(rna.new[,1], Esa.new[,1]),]
Set.sorted <- Set.new[match(rna.new[,1], Set.new[,1]),]
#preview
head(Gcn.sorted[,1:5])
head(Esa.sorted[,1:5])
head(Set.sorted[,1:5])
head(rna.new[,1:5])

```

```{r}
##Need to drop gene name
Gcn.sorted.cor <- dplyr::select(Gcn.sorted, -Gene.ID)
Esa.sorted.cor <- dplyr::select(Esa.sorted, -Gene.ID)
Set.sorted.cor <- dplyr::select(Set.sorted, -Gene.ID)
rna.new.cor <- dplyr::select(rna.new, -Gene.name)
corValues.Gcn <- sapply(1:nrow(Gcn.sorted.cor), function(i) cor(as.numeric(Gcn.sorted.cor[i,]), as.numeric(rna.new.cor[i,])))
corValues.Esa <- sapply(1:nrow(Esa.sorted.cor), function(i) cor(as.numeric(Esa.sorted.cor[i,]), as.numeric(rna.new.cor[i,])))
corValues.Set <- sapply(1:nrow(Set.sorted.cor), function(i) cor(as.numeric(Set.sorted.cor[i,]), as.numeric(rna.new.cor[i,])))
df.cors <- data.frame(corValues.Gcn, corValues.Esa, corValues.Set) 
head(df.cors)
```
```{r}
# But first lets append the gene names to the df.cors dataframe for more informative display
df.cors.w.names <- df.cors
row.names(df.cors.w.names) <- Gcn.sorted$Gene.ID #doesn't matter what set the names comes from - they are all the same 
head(df.cors.w.names)
```
```{r}
Gcn.indexes <- which(df.cors.w.names[,1] > .75)
Gcn.genes.above <- kuangtf.new[Gcn.indexes, 1]
Esa.indexes <- which(df.cors[,2] > .75)
Esa.genes.above <- kuangtf.new[Esa.indexes, 1]
Set.indexes <- which(df.cors[,3] > .75)
Set.genes.above <- kuangtf.new[Set.indexes, 1]
#print first 10 genes for each TF that meet threshold
head(df.cors.w.names[Gcn.indexes,1, drop = FALSE], 10) 
head(df.cors.w.names[Esa.indexes,2, drop = FALSE], 10) 
head(df.cors.w.names[Set.indexes,3, drop = FALSE], 10) 
```

```{r}
df.cors["gene"] = Set.sorted["Gene.ID"]
df.cors.single <- df.cors %>% 
  tidyr::gather(TF, corr, -gene) %>% # cast to long format
  dplyr::select(TF, gene, corr)
df.cors.single.trimmed <- filter(df.cors.single, corr > 0.75)

gcn <- filter(df.cors.single.trimmed, TF=="corValues.Gcn")
esa <- filter(df.cors.single.trimmed, TF=="corValues.Esa")
set <- filter(df.cors.single.trimmed, TF=="corValues.Set")
```


Now that we have the TF clusters from the previous homework, we can run a GO analysis on each cluster.
```{r}
# All GO Terms
gcn.go <- gprofiler(gcn, 
                          organism = "scerevisiae",
                          max_p_value = 0.05,  # set p-value cutoff
                          src_filter = "GO", 
                          hier_filtering = "moderate")
gcn.go
```
Cluster 1: Gcn5
Enriched Gene Ontology terms include:
BP: glycolytic process, ATP generation from ADP, hexose biosynthetic process
MF:	protein heterodimerization activity, NADP binding, glyceraldehyde-3-phosphate dehydrogenase (NAD(P)+) (phosphorylating) activity
CC: replication fork protection complex, nuclear nucleosome


```{r}
esa.go <- gprofiler(esa, 
                          organism = "scerevisiae",
                          max_p_value = 0.05,  # set p-value cutoff
                          src_filter = "GO",
                          hier_filtering = "moderate")
esa.go
```
Cluster 2: Esa1
Enriched Gene Ontology terms include:
BP: 
MF:	
CC: 

```{r}
set.go <- gprofiler(set, 
                          organism = "scerevisiae",
                          max_p_value = 0.05,  # set p-value cutoff
                          src_filter = "GO", 
                          hier_filtering = "moderate")
set.go
```
Cluster 3: Set1
Enriched terms include



We chose to look up a paper for the term with lowest p-value: **gluconeogenesis**. Gluconeogenesis is "a metabolic pathway that results in the generation of glucose from non-carbohydrate carbon substrates such as lactate, glycerol, and glucogenic amino acids".

Citation:
GimÃ©nez-Cassina, A., et. al. (2014). Regulation of hepatic energy metabolism and gluconeogenesis by BAD. Cell Metab, 19(2):272-84. doi: 10.1016/j.cmet.2013.12.001.


# Questions 3

Do you analyses shed light on how the specific biological processes of interest (identified in the GO analysis above) might be regulated or changed by the stress conditions that your data sets represent? 

Hierarchical Clustering Analysis

The biological processes of interest, identified by GO analysis above, such as the **monocarboxylic acid metabolic process** and **nicotinamide nucleotide metabolic process**, may contribute to the oscillations of gene expression under glucose limited conditions. The focal paper discusses the robust metabolic cycles associated with oscillations of gene expression that result from glucose-limited conditions in yeast. Since these terms came up in GO analysis and share similar genes of interest as the kuang data, it is possible that these processes are affected by glucose-limited conditions. 

TF Data Analysis

Our focal paper is about the effect of glucose-limited conditions on the metabolic cycle, and the gluconeogenesis paper is about a metabolic pathway that results in the generation of glucose. It makes sense that these papers would be interested in similar sets of genes. The focal paper discusses the robust metabolic cycles associated with oscillations of gene expression that result from glucose-limited conditions in yeast. It would be interesting to explore if gluconeogenesis is one of the metabolic cycles affected by glucose-limited conditions, and if it results in oscillations in the generation of glucose. If this is the case, then it could be a feedback loop where low glucose levels contribute to higher oscillations but higher glucose levels decrease the amount of oscillations. 

The other terms in the GO analysis for TF clusters do not seem to be as related to our main paper as terms produced by GO analysis of the hierarchical clusters. The terms in the GO analysis for TF clusters also have higher p-values. For this reason, we decided to focus on the GO terms from the hierarchical cluster analysis. 

# Conclusions, Biological Results, & Insights

TODO: bullet points
TODO: compare to the findings of the paper

- GO terms are related to the metabolic cycle and glucose
- The robust metabolic cycles associated with oscillations of gene expression that result from glucose-limited conditions (the findings from our focal paper) may be related to and affect biological/metabolic processes from the GO analysis such as  gluconeogenesis, monocarboxylic acid metabolic process, and nicotinamide nucleotide metabolic process

