---
title: "Group 10 TF HW"
author: "Group 10 members"
date: "March 30, 2017"
output:
  html_document: default
  html_notebook: default
---

# Group 10

Ryan St.Pierre (ras70)
Dakota Brinkman (dlb46)
Matt Olson (meo8)
Jeremy Schreck (jes85)

```{r setup, include=FALSE}
library(dplyr);
library("stringr");
library(ggplot2);
library(dplyr);
library(tidyr);
library(RColorBrewer);
library(dendextend);
library(gplots);
library(cluster);
library(igraph)
library(ggraph)
library(gridExtra)
library(grid)
```

## Question 1

The biological question we are trying to answer is how robust metabolic cycles within yeast cells respond under continuous, glucose limited conditions. Particuarly we aim to answer which clusters of genes control the process, which TF/chromatin modifies control their state, and how the interaction *between* clusters trough the transcription regulatory network contribute to the robust metabolic process. The two papers we were given try to correlate the chromatin states and this metabolic expression. After looking at what gene expression reveals about the function of the cell under these conditions, we explored the TFs that control the given clusters and how the TFs create complex functionality. Our biological question has not changed significantely, and we attempt understand the metabolic cycle within yeast cells with this current homework. However, we do aim to better tie the effect of the chromatin state, which is obviously tied to the chromatin modifiers, into our analysis and conclusions. We feel that our understanding of how the chromatin modifies contribute the process was lacking in the prior homework, which was a consequence of us conducting a gene clustering analysis and not a TF analysis.

Additionally we aim to derive more meaningful conclusions about which timepoints in the data contribute to the phases found in the paper (RC, OX, RB, RC). We wish to provide more specifity on the delay between these phases and potentially the mechanism driving them. We believe a transcription focused analysis will be instrumental in getting close in answering some of these questions

From our clustering analysis we were able to determine that there are ~10 clusters in the given expression data, a couple of which were extremely well correlated (temporally). Additionally, we were able to determine that several clusters display similar oscillatory patterns, with a slight phase (delay) difference. This suggests a "just-in-time supply chain" paradigm, in which components from specific processes" "become available in a highly correlated manner" [Abstract](http://www.nature.com/nsmb/journal/v21/n10/full/nsmb.2881.html#supplementary-information). This is well supported by the findings in the paper, which outlines several phases of the metabolic process: RC, OX, RB, RC. In this homework we seek to further investigate how the TF network contributes to this highly coordinated process. 

The TFs that we have been given, in the primary dataset, for this analysis are **Gcn5p, Esa1p and Set1p**. These are three critical chromatin modifiers, which means they alter the state of the chromatin, allowing other TFs to bind to the DNA, leading to gene expression. The format of this dataset is explained at depth in *question 2*. Additionally, we were given a broader TF dataset, which is investigated at the end of this analysis. 

## Question 2

We were given TF binding data for 3 TFs (Gcn, Esa, and Set) in the "kuang-2014-TFbinding.csv"" file. In order to analyze the format of the data we first imported it.

```{r}
kuangtf <- read.csv("./datasets/kuang-2014-TFbinding.csv")
time <- read.csv("./datasets/kuang-2014-TFbinding-timepts.csv")
kuangrna <- read.csv("./datasets/kuang-2014-rnaseq-expression.csv")   ##used later - included in the import section of the code for clarity
head(kuangtf)
head(time)
```

```{r}
dim(kuangtf)
dim(time)
```

From the output above it is clear that our data is **time series without knockout**. For each TF there are 14 columns, each of which correspond to a time in the time cvs data file. The rows of the dataset correspond to genes specifically. Each cell of the data gives the binding of the given TF and its time-period (according to the column) for a given gene (given by the column).

The rows of the dataset contain 1035 genes. We were curious why 1035 genes were given. After further investigation in the paper we found the following statement:

  "Among these, ~1,750 genes were bound by at least two modifiers, and 1,035 genes were bound by all three modifiers, thus suggesting that these enzymes were not necessary for expression of every cycling gene."
  [Landscape of occupancy of chromatin-modifying enzymes in the YMC](http://www.nature.com/nsmb/journal/v21/n10/full/nsmb.2881.html#supplementary-information)

The statement above clarifies which TFs we were given, those for which all 3 TFs did bind. 

### Threshold 

Given the fact we were given time series data, as opposed to p-value data, makes answering the question "What threshold did the paper use?" much more difficult. The three TFs given are already "known" to bind to the 1035 genes given (discussed above). This was determined through a spatiotemporal analysis. However, the paper is more concerned with trying to gain insight from analyzing the temporal attributes of the binding data. In more plain terms, the authors attempt to associate peak TF binding with a given genes expression. This type of analysis was done by directly comparing temproal data, and associating peaks, rather than chosing a threshold value. 

### Splitting up data and Reformatting

At this point we split the data, per TF and reformatted the data.

```{r}
Gcn <- kuangtf[,c(1:15)]
Esa <- kuangtf[,c(1,16:29)]
Set <- kuangtf[,c(1,30:43)]

Gcn.data <- as.data.frame(t(Gcn[,-1]))
colnames(Gcn.data) <- Gcn$Gene.ID
Gcn.time <- cbind(Gcn.data, time = time$hour)

Esa.data <- as.data.frame(t(Esa[,-1]))
colnames(Esa.data) <- Esa$Gene.ID
Esa.time <- cbind(Esa.data, time = time$hour)

Set.data <- as.data.frame(t(Set[,-1]))
colnames(Set.data) <- Set$Gene.ID
Set.time <- cbind(Set.data, time = time$hour)
```

### Plotting

TODO: talk about plots

```{R}
ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YAL012W, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YAL012W, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YAL012W, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YAL001C")

ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YBL033C, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YBL033C, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YBL033C, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YBL033C")

ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YBL039C, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YBL039C, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YBL039C, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YBL039C")

ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YBL053W, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YBL053W, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YBL053W, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YBL053W")

ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YBL087C, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YBL087C, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YBL087C, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YBL087C")

ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YER086W, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YER086W, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YER086W, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YER086W")

ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YJL045W, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YJL045W, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YJL045W, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YJL045W")

ggplot() + geom_line(data=Gcn.time, aes(x = time, y = YGL236C, color="Gcn")) + geom_line(data=Set.time, aes(x = time, y = YGL236C, color="Set")) + geom_line(data=Esa.time, aes(x = time, y = YGL236C, color="Esa")) + labs(x = "Time (hours)", y = "Binding of Gene", title = "YGL236C")
```

## Question 3

To analyze the genes that were regulated by a specific TF, we needed the RNA seq data. This data contained gene expression over 16 time points for all the genes. The first 14 time points matched the 14 time points from the TF data. Then we filtered the genes so all genes from the TF binding data had a matching gene from the RNA seq data. With this data, we created a correlation algorithm that determined how similar the expression and binding data was for a specific gene. We had to come up with this analysis completely on our own, and we are very happy with the results. We will talk later about how we determined the clusters of genes that each TF regulated.

We chose a correlation value of TODO to distinguish which TFs regulated which genes.

TODO: describe how we did this
TODO: how did we choose a correlation value cutoff
TODO: talk about basic network stuff (will discuss in more detail later)

#Correlation Set Up
```{r}

#get the genes in the tf and rna data
TF.genes <- unique(kuangtf[,1])
rna.genes <- unique(kuangrna[,1])

#get rid of duplicated rows in the rna data
rna.unique <- unique(kuangrna)

#filter the rna data for the tf genes
rna.Genes <- filter(rna.unique, rna.unique[,1] %in% TF.genes)

#get the unique genes from this rna data
#may not need to do this again but did it just to be safe
rna.genes.u <- unique(rna.Genes[,1])

#It appears that all of the genes present in the TF data are not present in the rna data
#So need to filter again to get rid of these genes
kuangtf.new <- filter(kuangtf, TF.genes %in% rna.genes.u)

#now remove last two columns so the times add up
rna.new <- rna.Genes[,1:15]

dim(rna.new)
dim(kuangtf.new)

#now extract the different TFs
Gcn.new <- kuangtf.new[,c(1:15)]
dim(Gcn.new)
Esa.new <- kuangtf.new[,c(1,16:29)]
dim(Esa.new)
Set.new <- kuangtf.new[,c(1,30:43)]
dim(Set.new)

#here we see the 69 genes that are in the TF data but not the rnaseq
length(which( ! TF.genes %in% rna.genes))

#sort so that the genes are in the same order
Gcn.sorted <- Gcn.new[match(rna.new[,1], Gcn.new[,1]),]
Esa.sorted <- Esa.new[match(rna.new[,1], Esa.new[,1]),]
Set.sorted <- Set.new[match(rna.new[,1], Set.new[,1]),]

head(Gcn.sorted)
head(Esa.sorted)
head(Set.sorted)
head(rna.new)
```

#Correlation
```{r}
##Need to drop gene name
Gcn.sorted.cor <- select(Gcn.sorted, -Gene.ID)
Esa.sorted.cor <- select(Esa.sorted, -Gene.ID)
Set.sorted.cor <- select(Set.sorted, -Gene.ID)
rna.new.cor <- select(rna.new, -Gene.name)
corValues.Gcn <- sapply(1:nrow(Gcn.sorted.cor), function(i) cor(as.numeric(Gcn.sorted.cor[i,]), as.numeric(rna.new.cor[i,])))
corValues.Esa <- sapply(1:nrow(Esa.sorted.cor), function(i) cor(as.numeric(Esa.sorted.cor[i,]), as.numeric(rna.new.cor[i,])))
corValues.Set <- sapply(1:nrow(Set.sorted.cor), function(i) cor(as.numeric(Set.sorted.cor[i,]), as.numeric(rna.new.cor[i,])))
df.cors <- data.frame(corValues.Gcn, corValues.Esa, corValues.Set) 
df.cors

Gcn.genes.above <- kuangtf.new[Gcn.indexes, 1]
Esa.indexes <- which(df.cors[,2] > .8)
Esa.genes.above <- kuangtf.new[Esa.indexes, 1]
Set.indexes <- which(df.cors[,3] > .8)
Set.genes.above <- kuangtf.new[Set.indexes, 1]
```
#Basic Clustering Analysis
```{r}
df.cors["gene"] = Set.sorted["Gene.ID"]
df.cors.single <- df.cors %>% 
  tidyr::gather(TF, corr, -gene) %>% # cast to long format
  dplyr::select(TF, gene, corr)
df.cors.single.trimmed <- filter(df.cors.single, corr > .7)

hits.per.tf <- df.cors.single.trimmed %>% 
               group_by(TF) %>% 
               summarise(ngenes = length(TF))

hits.per.tf
```

4. TODO: number of genes (need to decide on correlation value)

```{r}
df.cors.graph <- graph_from_data_frame(df.cors.single.trimmed)
vcount(df.cors.graph)
ecount(df.cors.graph)
tfs <- unique(df.cors.single.trimmed$TF)
V(df.cors.graph)$type <- V(df.cors.graph)$name %in% tfs
V(df.cors.graph)$shape  <- c("square", "circle")[V(df.cors.graph)$type+1]
V(df.cors.graph)$color  <- c("steelblue", "lightcoral")[V(df.cors.graph)$type+1]
count_components(df.cors.graph)
df.cors.components <- decompose(df.cors.graph)
lapply(df.cors.components, vcount)
lapply(df.cors.components, ecount)


plot(df.cors.components[[1]], 
     layout = layout.auto,
     vertex.label.cex=0.4,
     vertex.label.degree=-90, vertex.label.dist=1)

```

```{r}
df.cors.projection <- bipartite.projection(df.cors.graph)
gene.proj <- df.cors.projection[[1]]
tf.proj <- df.cors.projection[[2]]

tf.shared.cut <- delete.edges(tf.proj, E(tf.proj)[weight < 1])
plot(tf.shared.cut, vertex.label.cex=0.5, vertex.label.dist=0.5 )
```


```{r}
#TODO: fix (make better directed edge graph)
tfs.only.graph <- graph_from_data_frame(df.cors.single.trimmed, directed = TRUE)
l <- layout.circle(tfs.only.graph)

plot(tfs.only.graph, edge.arrow.size=.4, vertex.shape="none", vertex.label.cex=0.5, edge.curved=0.1, vertex.label.font=2, vertex.label.color="gray40", edge.color="gray85", layout=l)
```


5. See the network plots above. We will discuss these networks in 6.
6.
TODO: what do edges represent
TODO: what do nodes represent
TODO: explain directed and undirected
TODO: how many TFs/ how many target genes
7.
TODO: out degree
TODO: in degree
8.
TODO: synthesize info
TODO: target genes direct targets of TF? indirect targets? How did we determine this?
TODO: functions of target genes?
TODO: any target genes a TF?


#TF vs. RNAseq Analysis
```{r}
Gcn.indexes <- which(df.cors[,1] > .8)
Gcn.genes.above <- kuangtf.new[Gcn.indexes, 1]
Esa.indexes <- which(df.cors[,2] > .8)
Esa.genes.above <- kuangtf.new[Esa.indexes, 1]
Set.indexes <- which(df.cors[,3] > .8)
Set.genes.above <- kuangtf.new[Set.indexes, 1]
```
We decided to use a cutoff of .8 for our correlation value.

```{r}
timechar <- c(
"94.95",
"95.31666667",
"95.41666667",
"95.47222222",
"95.59722222",
"95.74166667",
"95.98333333",
"96.24166667",
"96.35555556",
"96.55833333",
"97.05833333",
"97.55833333",
"98.05833333",
"98.56666667")

time <- as.numeric(timechar)
```

Let's analyze for the Gcn5 TF first
```{r}
Gcn.genes.above.plot <- as.character(Gcn.genes.above)
Gcn.above <- filter(Gcn.new, Gene.ID %in% Gcn.genes.above.plot)
Gcn.t <- t(Gcn.above)
Gcn.t <- Gcn.t[2:nrow(Gcn.t),]
length(Gcn.genes.above.plot)
colnames(Gcn.t) <- as.vector(Gcn.genes.above.plot)
Gcn.t.time <- cbind(Gcn.t, time = time)

Gcn.long <- gather(data.frame(Gcn.t.time), gene, binding, -time)
Gcn.long$time <- as.numeric(as.character(Gcn.long$time))
Gcn.long$binding <- as.numeric(as.character(Gcn.long$binding))
Gcn.long %>% filter(gene %in% Gcn.genes.above.plot) %>%
  ggplot(aes(x=time, y=binding, color=gene)) + ggtitle("Gcn5 TFbinding for genes with cor > .8") + geom_line()

Gcn.rna <- filter(rna.new, Gene.name %in% Gcn.genes.above.plot)
Gcn.rna.t <- t(Gcn.rna)
newGenes <- Gcn.rna.t[1,]
Gcn.rna.t2 <- Gcn.rna.t[2:nrow(Gcn.rna.t),]
colnames(Gcn.rna.t2) <- as.vector(newGenes)
Gcn.rna.time <- cbind(Gcn.rna.t2, time = time)
Gcn.rna.long <- gather(data.frame(Gcn.rna.time), gene, expression, -time)
Gcn.rna.long$time <- as.numeric(as.character(Gcn.rna.long$time))
Gcn.rna.long$expression <- as.numeric(as.character(Gcn.rna.long$expression))

Gcn.rna.long %>% filter(gene %in% Gcn.genes.above.plot) %>%
  ggplot(aes(x=time, y=expression, color=gene)) + ggtitle("Gcn5 RNAseq for genes with cor > .8") + geom_line()

gene.of.interest <- Gcn.genes.above.plot[1]
p1 <- Gcn.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="red")
p2 <- Gcn.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="red")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Gcn.genes.above.plot[1]), gp=gpar(fontface="bold")))

gene.of.interest <- Gcn.genes.above.plot[2]
p1 <- Gcn.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="red")
p2 <- Gcn.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="red")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Gcn.genes.above.plot[2]), gp=gpar(fontface="bold")))


gene.of.interest <- Gcn.genes.above.plot[3]
p1 <- Gcn.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="red")
p2 <- Gcn.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="red")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Gcn.genes.above.plot[3]), gp=gpar(fontface="bold")))
```

Now for Esa1
```{r}
Esa.genes.above.plot <- as.character(Esa.genes.above)
Esa.above <- filter(Esa.new, Gene.ID %in% Esa.genes.above.plot)
Esa.t <- t(Esa.above)
Esa.t <- Esa.t[2:nrow(Esa.t),]
length(Esa.genes.above.plot)
colnames(Esa.t) <- as.vector(Esa.genes.above.plot)
Esa.t.time <- cbind(Esa.t, time = time)
 
Esa.long <- gather(data.frame(Esa.t.time), gene, binding, -time)
Esa.long$time <- as.numeric(as.character(Esa.long$time))
Esa.long$binding <- as.numeric(as.character(Esa.long$binding))
Esa.long %>% filter(gene %in% Esa.genes.above.plot) %>%
  ggplot(aes(x=time, y=binding, color=gene)) + ggtitle("Esa1 TFbinding for genes with cor > .8") + geom_line()

Esa.rna <- filter(rna.new, Gene.name %in% Esa.genes.above.plot)
Esa.rna.t <- t(Esa.rna)
newGenes <- Esa.rna.t[1,]
Esa.rna.t2 <- Esa.rna.t[2:nrow(Esa.rna.t),]
colnames(Esa.rna.t2) <- as.vector(newGenes)
Esa.rna.time <- cbind(Esa.rna.t2, time = time)
Esa.rna.long <- gather(data.frame(Esa.rna.time), gene, expression, -time)
Esa.rna.long$time <- as.numeric(as.character(Esa.rna.long$time))
Esa.rna.long$expression <- as.numeric(as.character(Esa.rna.long$expression))

Esa.rna.long %>% filter(gene %in% Esa.genes.above.plot) %>%
  ggplot(aes(x=time, y=expression, color=gene)) + ggtitle("Esa1 TFbinding for genes with cor > .8") + geom_line()

gene.of.interest <- Esa.genes.above.plot[1]
p1 <- Esa.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="blue")
p2 <- Esa.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="blue")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Esa.genes.above.plot[1]), gp=gpar(fontface="bold")))

gene.of.interest <- Esa.genes.above.plot[2]
p1 <- Esa.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="blue")
p2 <- Esa.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="blue")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Esa.genes.above.plot[2]), gp=gpar(fontface="bold")))


gene.of.interest <- Esa.genes.above.plot[3]
p1 <- Esa.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="blue")
p2 <- Esa.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="blue")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Esa.genes.above.plot[3]), gp=gpar(fontface="bold")))
```

And finally for Set1
```{r}
Set.genes.above.plot <- as.character(Set.genes.above)
Set.above <- filter(Set.new, Gene.ID %in% Set.genes.above.plot)
Set.t <- t(Set.above)
Set.t <- Set.t[2:nrow(Set.t),]
length(Set.genes.above.plot)
colnames(Set.t) <- as.vector(Set.genes.above.plot)
Set.t.time <- cbind(Set.t, time = time)
 
Set.long <- gather(data.frame(Set.t.time), gene, binding, -time)
Set.long$time <- as.numeric(as.character(Set.long$time))
Set.long$binding <- as.numeric(as.character(Set.long$binding))
Set.long %>% filter(gene %in% Set.genes.above.plot) %>%
  ggplot(aes(x=time, y=binding, color=gene)) + ggtitle("Set1 TFbinding for genes with cor > .8") + geom_line()

Set.rna <- filter(rna.new, Gene.name %in% Set.genes.above.plot)
Set.rna.t <- t(Set.rna)
newGenes <- Set.rna.t[1,]
Set.rna.t2 <- Set.rna.t[2:nrow(Set.rna.t),]
colnames(Set.rna.t2) <- as.vector(newGenes)
Set.rna.time <- cbind(Set.rna.t2, time = time)
Set.rna.long <- gather(data.frame(Set.rna.time), gene, expression, -time)
Set.rna.long$time <- as.numeric(as.character(Set.rna.long$time))
Set.rna.long$expression <- as.numeric(as.character(Set.rna.long$expression))

Set.rna.long %>% filter(gene %in% Set.genes.above.plot) %>%
  ggplot(aes(x=time, y=expression, color=gene)) + ggtitle("Set1 TFbinding for genes with cor > .8") + geom_line()

gene.of.interest <- Set.genes.above.plot[1]
p1 <- Set.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="green")
p2 <- Set.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="green")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Set.genes.above.plot[1]), gp=gpar(fontface="bold")))

gene.of.interest <- Set.genes.above.plot[4]
p1 <- Set.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="green")
p2 <- Set.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="green")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Set.genes.above.plot[4]), gp=gpar(fontface="bold")))


gene.of.interest <- Set.genes.above.plot[3]
p1 <- Set.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=binding)) + ggtitle("TFbinding") + geom_line(colour="green")
p2 <- Set.rna.long %>% filter(gene %in% gene.of.interest) %>%
  ggplot(aes(x=time, y=expression)) + ggtitle("RNAseq Expression") + geom_line(colour="green")
grid.arrange(p1, p2, ncol=2, top=textGrob(as.character(Set.genes.above.plot[3]), gp=gpar(fontface="bold")))
```


9.

TODO: talk about TF binding for TF cluster plot
TODO: talk about RNA seq for TF cluster plot
TODO: talk about TF binding and RNA seq plot
TODO: TF clusters vs gene expression clusters? Do they fall into same sets?

```{r}
tf.ypd <- read.csv(file="./datasets/harbison-2004-TFbinding-YPD.txt", sep="\t",header=T)
tf.other <- read.csv(file="./datasets/harbison-2004-TFbinding-other.txt", sep="\t",header=T)
head(tf.ypd)
dim(tf.ypd)
head(tf.other)
dim(tf.other)
```

```{r}
tf.ypd.prepped <- na.omit(tf.ypd)
dim(tf.ypd.prepped)
```

Conclusion
TODO: talk briefly about biological questions and major findings
